### DISCUSSION TITLE
How to create a substructure pattern with disconnected fragments?

### DISCUSSION THREAD

**Discussion by GLPG-GT** [rdkit/rdkit#6359] [Q&A] [answered]
Apologies in advance for the long introduction, the topic is not so simple.

Imagine you have a molecule, and you want to create analogues of it by changing some part(s) of it, _while keeping the rest of the molecule 'fixed'_.  

Example:
```
from rdkit import Chem
from rdkit.Chem.Draw import IPythonConsole
IPythonConsole.drawOptions.addAtomIndices = True
IPythonConsole.molSize = 450,300

m = Chem.MolFromSmiles('N2(C(C1=CC=CC=C1)CCC(=O)N)C=CC=N2')
m
```
![download](https://github.com/rdkit/rdkit/assets/96192927/fe348087-008e-4911-82c0-a73323080060)

You specify the indices of the heavy atoms to change:

`selected_atoms = [1, 8, 9]`

You identify the implicit H's to change:
```
m_allHs = Chem.AddHs(m)
m_allHs
```
![download](https://github.com/rdkit/rdkit/assets/96192927/3723ad63-8d38-4abd-b451-1a63fad91925)
```
implicit_Hs_to_change = []
for a in m_allHs.GetAtoms() :
    if (a.GetAtomicNum() != 1) & (a.GetIdx() in selected_atoms) :
        for att in a.GetNeighbors() :
            if att.GetAtomicNum() == 1 :
                implicit_Hs_to_change.append(att.GetIdx())
implicit_Hs_to_change
# [17, 23, 24, 25, 26]
```
Now, suppose you have some software that does the modifications you want, like replacing atoms, adding atoms, etc.  
The software takes as input the molecule (with all H's) and the list of atoms that must NOT be changed. It's easy to find them by exclusion:

`atoms_to_keep_fixed = [a.GetIdx() for a in m_allHs.GetAtoms() if ((a.GetIdx() not in selected_atoms) & (a.GetIdx() not in implicit_Hs_to_change))]`

You may obtain this molecule as valid output:
```
m_new = Chem.MolFromSmiles('N2(C(C1=CC=CC=C1)OCCC(=O)N)C=CC=N2')
m_new
```
![download](https://github.com/rdkit/rdkit/assets/96192927/fd1def8e-659e-402a-aafc-4fb4a2856bb0)

The software did a good job, because it kept the moieties that you did not select for changing.  
But the 'skeleton' of the molecule no longer matches your initial one.
So if you wanted to _further_ modify this new molecule, _while still keeping the 'fixed' atoms as in your parent molecule_, you would have to re-match it to the fixed parts of the original molecule.

This raises two questions: 1) what happens if I remove the atoms to change from the parent molecule, to make a pattern out of it? (in this case this would split a single molecule into fragments); 2) will rdkit be able to match the new molecule to a pattern with multiple fragments?

The answer to Question 2, I found out to be 'yes', by sketching myself manually the fragmented pattern, and matching it:
```
ps = Chem.SmilesParserParams()
ps.removeHs = False
p = Chem.MolFromSmiles('N1C([H])=C([H])C([H])=N1.C1=C([H])C([H])=C([H])C([H])=C1[H].C(=O)N([H])[H]', ps)
p
```
![download](https://github.com/rdkit/rdkit/assets/96192927/1469c811-6eca-4620-949d-2758bea8eb47)
```
Chem.AddHs(m_new).GetSubstructMatch(p)
# (0,
#  14,
#  30,
#  15,
#  31,
#  ...
```
Question 1 is linked to [this post](https://github.com/rdkit/rdkit/discussions/4685) and perhaps also [this post](https://github.com/rdkit/rdkit/discussions/5789), I guess.

First naive thing I tried was to just remove the atoms and see:
```
mrw = Chem.RWMol(m_allHs)
mrw.BeginBatchEdit()
for a in mrw.GetAtoms():
    aid = a.GetIdx()
    if aid not in atoms_to_keep_fixed :
        mrw.RemoveAtom(aid)
mrw.CommitBatchEdit()
Chem.SanitizeMol(mrw)
mrw
```
which, not unexpectedly, resulted in big red flashing errors:

`KekulizeException: Can't kekulize mol.  Unkekulized atoms: 0 10 11 12 13`

and obviously did not result in a matching pattern:
```
Chem.AddHs(m_new).GetSubstructMatch(mrw)
# ()
```
I was actually unsure what to do next, because I did not find any post circumventing the issue described in the linked ones, and I could not use a reaction (could I?).

So just out of curiosity I tried this (going via SMARTS):
```
psm = Chem.MolToSmarts(m_allHs)
psmm = Chem.MolFromSmarts(psm)
atoms_to_keep_fixed_psmm = [aid for i,aid in enumerate(list(psmm.GetSubstructMatch(m_allHs))) if i in atoms_to_keep_fixed]
psmmrw = Chem.RWMol(psmm)
psmmrw.BeginBatchEdit()
for a in psmmrw.GetAtoms():
    aid = a.GetIdx()
    if aid not in atoms_to_keep_fixed_psmm :
        psmmrw.RemoveAtom(aid)
psmmrw.CommitBatchEdit()
Chem.SanitizeMol(psmmrw)
psmmrw
```
![download](https://github.com/rdkit/rdkit/assets/96192927/8c297c08-718c-4303-b891-5a4612a25689)

This gave no error, and it also worked in matching the new molecule:
```
Chem.AddHs(m_new).GetSubstructMatch(psmmrw)
# (0,
#  2,
#  3,
#  4,
#  5,
#  6,
#  7,
#  23,
#  22,
#  21,
#  20,
#  19,
#  11,
#  12,
#  13,
#  28,
#  29,
#  14,
#  15,
#  16,
#  17,
#  32,
#  31,
#  30)
```
I am aware that this way of matching may go awry if some substructure that was in the original pattern is inserted in the 'changed' part of the molecule. That's OK for now.

But my main questions are on the technical aspect of the method I used:
**Do you think the above method via SMARTS is legitimate?
Do you foresee any issues or malfunctions?
Would you recommend a different/better approach?**

**Comment by DavidACosgrove** [comment]
If I've understood what you're doing correctly, you might find it easier to
use atom properties.  You can set an arbitrary property on an atom that
says whether it is to be kept or not.  Any modified copy of the molecule
will retain the atom properties for the next round.  Your modification code
can then read the atom properties before deciding what to do with that
atom.  There will be no need to do any substructure searching, so it may
also be quicker.


On Wed, May 10, 2023 at 3:15 PM GLPG-GT ***@***.***> wrote:

> Apologies in advance for the long introduction, the topic is not so simple.
>
> Imagine you have a molecule, and you want to create analogues of it by
> changing some part(s) of it, *while keeping the rest of the molecule
> 'fixed'*.
>
> Example:
>
> from rdkit import Chem
> from rdkit.Chem.Draw import IPythonConsole
> IPythonConsole.drawOptions.addAtomIndices = True
> IPythonConsole.molSize = 450,300
>
> m = Chem.MolFromSmiles('N2(C(C1=CC=CC=C1)CCC(=O)N)C=CC=N2')
> m
>
> [image: download]
> <https://user-images.githubusercontent.com/96192927/237406356-fe348087-008e-4911-82c0-a73323080060.png>
>
> You specify the indices of the heavy atoms to change:
>
> selected_atoms = [7, 14, 15]
>
> You identify the implicit H's to change:
>
> m_allHs = Chem.AddHs(m)
> m_allHs
>
> [image: download]
> <https://user-images.githubusercontent.com/96192927/237406949-3723ad63-8d38-4abd-b451-1a63fad91925.png>
>
> implicit_Hs_to_change = []
> for a in m_allHs.GetAtoms() :
>     if (a.GetAtomicNum() != 1) & (a.GetIdx() in selected_atoms) :
>         for att in a.GetNeighbors() :
>             if att.GetAtomicNum() == 1 :
>                 implicit_Hs_to_change.append(att.GetIdx())
> implicit_Hs_to_change
> # [17, 23, 24, 25, 26]
>
> Now, suppose you have some software that does the modifications you want,
> like replacing atoms, adding atoms, etc.
> The software takes as input the molecule (with all H's) and the list of
> atoms that must NOT be changed. It's easy to find them by exclusion:
>
> atoms_to_keep_fixed = [a.GetIdx() for a in m_allHs.GetAtoms() if
> ((a.GetIdx() not in selected_atoms) & (a.GetIdx() not in
> implicit_Hs_to_change))]
>
> You may obtain this molecule as valid output:
>
> m_new = Chem.MolFromSmiles('N2(C(C1=CC=CC=C1)OCCC(=O)N)C=CC=N2')
> m_new
>
> [image: download]
> <https://user-images.githubusercontent.com/96192927/237408003-fd1def8e-659e-402a-aafc-4fb4a2856bb0.png>
>
> The software did a good job, because it kept the moieties that you did not
> select for changing.
> But the 'skeleton' of the molecule no longer matches your initial one.
> So if you wanted to *further* modify this new molecule, *while still
> keeping the 'fixed' atoms as in your parent molecule*, you would have to
> re-match it to the fixed parts of the original molecule.
>
> This raises two questions: 1) what happens if I remove the atoms to change
> from the parent molecule, to make a pattern out of it? (in this case this
> would split a single molecule into fragments); 2) will rdkit be able to
> match the new molecule to a pattern with multiple fragments?
>
> The answer to Question 2, I found out to be 'yes', by sketching myself
> manually the fragmented pattern, and matching it:
>
> ps = Chem.SmilesParserParams()
> ps.removeHs = False
> p = Chem.MolFromSmiles('N1C([H])=C([H])C([H])=N1.C1=C([H])C([H])=C([H])C([H])=C1[H].C(=O)N([H])[H]', ps)
> p
>
> [image: download]
> <https://user-images.githubusercontent.com/96192927/237413625-1469c811-6eca-4620-949d-2758bea8eb47.png>
>
> Chem.AddHs(m_new).GetSubstructMatch(p)
> # (0,
> #  14,
> #  30,
> #  15,
> #  31,
> #  ...
>
> Question 1 is linked to this post
> <https://github.com/rdkit/rdkit/discussions/4685> and perhaps also this
> post <https://github.com/rdkit/rdkit/discussions/5789>, I guess.
>
> First naive thing I tried was to just remove the atoms and see:
>
> mrw = Chem.RWMol(m_allHs)
> mrw.BeginBatchEdit()
> for a in mrw.GetAtoms():
>     aid = a.GetIdx()
>     if aid not in atoms_to_keep_fixed :
>         mrw.RemoveAtom(aid)
> mrw.CommitBatchEdit()
> Chem.SanitizeMol(mrw)
> mrw
>
> which, not unexpectedly, resulted in big red flashing errors:
>
> KekulizeException: Can't kekulize mol. Unkekulized atoms: 0 10 11 12 13
>
> and obviously did not result in a matching pattern:
>
> Chem.AddHs(m_new).GetSubstructMatch(mrw)
> # ()
>
> I was actually unsure what to do next, because I did not find any post
> circumventing the issue described in the linked ones, and I could not use a
> reaction (could I?).
>
> So just out of curiosity I tried this (going via SMARTS):
>
> psm = Chem.MolToSmarts(m_allHs)
> psmm = Chem.MolFromSmarts(psm)
> atoms_to_keep_fixed_psmm = [aid for i,aid in enumerate(list(psmm.GetSubstructMatch(m_allHs))) if i in atoms_to_keep_fixed]
> psmmrw = Chem.RWMol(psmm)
> psmmrw.BeginBatchEdit()
> for a in psmmrw.GetAtoms():
>     aid = a.GetIdx()
>     if aid not in atoms_to_keep_fixed_psmm :
>         psmmrw.RemoveAtom(aid)
> psmmrw.CommitBatchEdit()
> Chem.SanitizeMol(psmmrw)
> psmmrw
>
> [image: download]
> <https://user-images.githubusercontent.com/96192927/237415866-8c297c08-718c-4303-b891-5a4612a25689.png>
>
> This gave no error, and it also worked in matching the new molecule:
>
> Chem.AddHs(m_new).GetSubstructMatch(psmmrw)
> # (0,
> #  2,
> #  3,
> #  4,
> #  5,
> #  6,
> #  7,
> #  23,
> #  22,
> #  21,
> #  20,
> #  19,
> #  11,
> #  12,
> #  13,
> #  28,
> #  29,
> #  14,
> #  15,
> #  16,
> #  17,
> #  32,
> #  31,
> #  30)
>
> I am aware that this way of matching may go awry if some substructure that
> was in the original pattern is inserted in the 'changed' part of the
> molecule. That's OK for now.
>
> But my main questions are on the technical aspect of the method I used:
>
>
> *Do you think the above method via SMARTS is legitimate? Do you foresee
> any issues or malfunctions? Would you recommend a different/better
> approach?*
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/6359>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FTMEO3RIIX7HIIFRDLXFOPJZANCNFSM6AAAAAAX4Z3UR4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by GLPG-GT** [reply]
Thank you David, this is interesting.
I am not sufficiently good with rdkit to see exactly how I would do that, though.
In practice, how do you suggest I could find the atom indices in `Chem.AddHs(m_new)` that match the `atoms_to_keep_fixed` in `m_allHs`?

**Comment by DavidACosgrove** [comment, marked as answer]
Something like this, maybe?
```
for a in m_allHs.GetAtoms():
    if a.GetIdx() not in selected_atoms and a.GetIdx() not in
implicit_Hs_to_change:
        a.SetIntProp("DontChange", 1)

other_mol = Chem.Mol(m_allHs)
for oa in other_mol.GetAtoms():
    if oa.HasProp("DontChange"):
        print(f'Not touching {oa.GetIdx()}')
    else:
        print(f'{oa.GetIdx()} is ok to edit')```

On Wed, May 10, 2023 at 4:07 PM GLPG-GT ***@***.***> wrote:

> Thank you David, this is interesting.
> I am not sufficiently good with rdkit to see exactly how I would do that,
> though.
> In practice, how do you suggest I could find the atom indices in
> Chem.AddHs(m_new) that match the atoms_to_keep_fixed in m_allHs?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/6359#discussioncomment-5862156>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FU24437CVIWZCX6Y6TXFOVMPANCNFSM6AAAAAAX4Z3UR4>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by GLPG-GT** [reply]
OK I see, thanks.
Unfortunately this does not work, if I understood correctly what you meant.
I need to identify the atoms in `Chem.AddHs(m_new)` that match the ones in `m_allHs`.
And by definition, `Chem.AddHs(m_new)` can have no pre-set "DontChange" property attached to any atoms, because the process that makes modifications to `m_allHs` returns SMILES as output, thus removing any special atom properties.
Sorry, I did not provide details of the modification, to avoid making my post even longer than it was.
But yes, if the modifications were made on molecular objects and we could retain atom properties, this would really work very well.

**Reply by GLPG-GT** [reply]
Actually, I just found that the modification software can also return molecules, not just SMILES, and it retains the atom properties. So your solution works!
Thank you very much indeed!
